src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,178 @@
+    AwaitReply \<Rightarrow> Idle \<comment> \<open>only for this initial implementation\<close>
+  )"
+
+definition state_machine where
+  [simp]: "state_machine = \<lparr>
+    initial_state = Idle,
+    agent_in_state = agent_in_state',
+    can_finish_in_state = can_finish_in_state',
+    next_state = next_state'
+  \<rparr>"
+
+sublocale chain_sync \<subseteq> protocol_state_machine \<open>state_machine\<close> .
+
+subsection \<open>Programs\<close>
+
+datatype phase =
+  is_intersection_finding: IntersectionFinding |
+  is_items_update: ItemsUpdate
+
+definition roll_back_to :: "'i list \<Rightarrow> 'p \<Rightarrow> ('i \<Rightarrow> 'p) \<Rightarrow> 'i list" where
+  [simp]: "roll_back_to \<I> p \<psi> = (THE \<I>\<^sub>1. \<exists>\<I>\<^sub>2. \<I> = \<I>\<^sub>1 @ \<I>\<^sub>2 \<and> \<psi> (last \<I>\<^sub>1) = p)"
+
+corec client_program where
+  "client_program ph \<kappa> \<psi> \<I> = (
+    case ph of
+      IntersectionFinding \<Rightarrow>
+        \<up> Cont (FindIntersect (\<kappa> \<I>));
+        \<down> M; (partial_case M of
+          Cont IntersectNotFound \<Rightarrow>
+            \<up> Done;
+            \<bottom> |
+          Cont (IntersectFound _) \<Rightarrow>
+            client_program ItemsUpdate \<kappa> \<psi> \<I>
+        ) |
+      ItemsUpdate \<Rightarrow>
+        \<up> Cont RequestNext;
+        \<down> M; (partial_case M of
+          Cont (RollForward i) \<Rightarrow>
+            client_program ph \<kappa> \<psi> (\<I> @ [i]) |
+          Cont (RollBackward p) \<Rightarrow>
+            client_program ph \<kappa> \<psi> (roll_back_to \<I> p \<psi>) |
+          Cont AwaitReply \<Rightarrow> \<comment> \<open>client is up to date\<close>
+            \<up> Done;
+            \<bottom>
+        )
+    )"
+
+definition index_of :: "'p \<Rightarrow> 'i list \<Rightarrow> ('i \<Rightarrow> 'p) \<Rightarrow> nat" where
+  [simp]: "index_of p \<I> \<psi> = (THE k. \<psi> (\<I> ! k) = p)"
+
+corec server_program where
+  "server_program rp mrb \<rho> \<psi> \<I> =
+    \<down> M; (partial_case M of
+      Done \<Rightarrow>
+        \<bottom> |
+      Cont (FindIntersect ps) \<Rightarrow> (
+        case \<rho> \<I> ps of
+          None \<Rightarrow>
+            \<up> Cont IntersectNotFound;
+            server_program rp mrb \<rho> \<psi> \<I> |
+          Some p \<Rightarrow>
+            \<up> Cont (IntersectFound p);
+            server_program (index_of p \<I> \<psi>) True \<rho> \<psi> \<I>
+      ) |
+      Cont RequestNext \<Rightarrow>
+        if mrb then
+          \<up> Cont (RollBackward (\<psi> (\<I> ! rp)));
+          server_program (Suc rp) False \<rho> \<psi> \<I>
+        else
+          if rp < length \<I> then
+            \<up> Cont (RollForward (\<I> ! rp));
+            server_program (Suc rp) mrb \<rho> \<psi> \<I>
+          else \<comment> \<open>client is up to date\<close>
+            \<up> Cont AwaitReply;
+            server_program rp mrb \<rho> \<psi> \<I>
+    )"
```

Comment:
>Several comments made for the client program also apply to the server program and should be taken into account also here.

Fixed in d27a93ddbb76a0074b18ad2f6933571c3d845ba9, c78098b5c4776e946b1b6288e018ee86aab31c5b, and a0ef9d83c6698ed03969451ca5362f8dff145b01. Please check.

Reply:
> > Several comments made for the client program also apply to the server program and should be taken into account also here.
> 
> Fixed in [d27a93d](https://github.com/input-output-hk/ouroboros-high-assurance/commit/d27a93ddbb76a0074b18ad2f6933571c3d845ba9), [c78098b](https://github.com/input-output-hk/ouroboros-high-assurance/commit/c78098b5c4776e946b1b6288e018ee86aab31c5b), and [a0ef9d8](https://github.com/input-output-hk/ouroboros-high-assurance/commit/a0ef9d83c6698ed03969451ca5362f8dff145b01). Please check.

Regarding c78098b5c4776e946b1b6288e018ee86aab31c5b, I’m not so sure meanwhile whether my approach of starting `case` expressions on the same line is so good in this case. With `case φ` in `client_program` we have a case distinction for a function argument right at the start of the right-hand side of the defining equation. Here, however, we are within an alternative of another `case` expression and our scrutinee is a complex term. I think that we should handle the case expression that starts with `case first_intersection_point ψ ps 𝒞` like a normal expression. This means going back to the previous layout with two exceptions:
* The opening parenthesis before the `case` expression should not come after `⇒` but be at the start of the following line just before the `case`.
* The closing parenthesis after the `case` expression should be indented one level more.

Besides that, there is [another comment of mine regarding `client_program`](https://github.com/input-output-hk/ouroboros-high-assurance/pull/74#discussion_r1333661672) that should be analogously taken into account also for `server_program`.

Reply:
> * The closing parenthesis after the `case` expression should be indented one level more.

I'm not completely sure about the layout that you propose. Is it the following one?
```
      Cont (FindIntersect ps) ⇒
        (case first_intersection_point ψ ps 𝒞 of
          None ⇒
            ↑ Cont IntersectNotFound;
            server_program ψ 𝒞 rp mrb |
          Some p ⇒
            ↑ Cont (IntersectFound p);
            server_program ψ 𝒞 (index ψ p 𝒞) True
        ) |
```

Reply:
> > * The closing parenthesis after the `case` expression should be indented one level more.
> 
> I'm not completely sure about the layout that you propose. Is it the following one?
> 
> ```
>       Cont (FindIntersect ps) ⇒
>         (case first_intersection_point ψ ps 𝒞 of
>           None ⇒
>             ↑ Cont IntersectNotFound;
>             server_program ψ 𝒞 rp mrb |
>           Some p ⇒
>             ↑ Cont (IntersectFound p);
>             server_program ψ 𝒞 (index ψ p 𝒞) True
>         ) |
> ```

Yes, this is the one. :+1:

Reply:
> > Besides that, there is [another comment of mine regarding `client_program`](https://github.com/input-output-hk/ouroboros-high-assurance/pull/74#discussion_r1333661672) that should be analogously taken into account also for `server_program`.
> 
> Fixed in [2f8c216](https://github.com/input-output-hk/ouroboros-high-assurance/commit/2f8c21659980c65eb09208699507d378e28e5525), please check.

Hmm, what about changing `r` to `b` (“roll *backward*) in order for it to not be mistaken for a *read* pointer?

The use of `k` is actually fine in my opinion.

Reply:
Please don’t forget to take my second, long comment in this thread into account.

Reply:
> Hmm, what about changing `r` to `b` (“roll _backward_”) in order for it to not be mistaken for a _read_ pointer?

Fixed in 45cf3509b10ad4d26446fe675f91c2688fcefbdd, please check.

Reply:
If we include genesis items in chains as described in [a comment above](https://github.com/input-output-hk/ouroboros-high-assurance/pull/74#discussion_r1337628029), we can actually stick to read pointers, as it would then be the genesis items that have index 0 (and read pointers cannot point *before* genesis items, as genesis items are never transmitted). That said, if we stick to read pointers, we should actually make sure that our read pointers are indeed in every situation what *The Shelley Networking Protocol* calls read pointers: pointers to items that are directly *before* items to be transmitted next.